src/core/api/health.go

```diff
@@ -0,0 +1,273 @@
+
+// HealthAPI handles the request for "/api/health"
+type HealthAPI struct {
+	BaseController
+}
+
+// CheckHealth checks the health of system
+func (h *HealthAPI) CheckHealth() {
+	healthy := true
+	components := []map[string]string{}
+	for name, checker := range healthCheckerRegistry {
+		err := checker.Check()
+		componentHealthy := err == nil
+		componentStatus := map[string]string{
+			"name":   name,
+			"status": healthyStatus(componentHealthy),
+		}
+		if !componentHealthy {
+			componentStatus["error"] = err.Error()
+			healthy = false
+		}
+		components = append(components, componentStatus)
+	}
+	status := map[string]interface{}{}
+	status["status"] = healthyStatus(healthy)
+	status["components"] = components
+	if !healthy {
+		log.Debugf("unhealthy system status: %v", status)
+	}
+	h.WriteJSONData(status)
+}
+
+func healthyStatus(healthy bool) string {
+	if healthy {
+		return "healthy"
+	}
+	return "unhealthy"
+}
+
+// HTTPStatusCodeHealthChecker implements a Checker to check that the HTTP status code
+// returned matches the expected one
+func HTTPStatusCodeHealthChecker(method string, url string, header http.Header,
+	timeout time.Duration, statusCode int) health.Checker {
+	return health.CheckFunc(func() error {
+		req, err := http.NewRequest(method, url, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create request: %v", err)
+		}
+		for key, values := range header {
+			for _, value := range values {
+				req.Header.Add(key, value)
+			}
+		}
+
+		client := httputil.NewClient(&http.Client{
+			Timeout: timeout,
+		})
+		resp, err := client.Do(req)
+		if err != nil {
+			return fmt.Errorf("failed to check health: %v", err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != statusCode {
+			data, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				log.Debugf("failed to read response body: %v", err)
+			}
+			return fmt.Errorf("received unexpected status code: %d %s", resp.StatusCode, string(data))
+		}
+
+		return nil
+	})
+}
+
+type updater struct {
+	sync.Mutex
+	status error
+}
+
+func (u *updater) Check() error {
+	u.Lock()
+	defer u.Unlock()
+
+	return u.status
+}
+
+func (u *updater) update(status error) {
+	u.Lock()
+	defer u.Unlock()
+
+	u.status = status
+}
+
+// PeriodicHealthChecker implements a Checker to check status periodically
+func PeriodicHealthChecker(checker health.Checker, period time.Duration) health.Checker {
+	u := &updater{
+		// init the "status" as "unknown status" error to avoid returning nil error(which means healthy)
+		// before the first health check request finished
+		status: errors.New("unknown status"),
+	}
+
+	go func() {
+		ticker := time.NewTicker(period)
+		for {
+			u.update(checker.Check())
+			<-ticker.C
+		}
+	}()
+
+	return u
+}
+
+func coreHealthChecker() health.Checker {
+	return health.CheckFunc(func() error {
+		return nil
+	})
+}
+
+func portalHealthChecker() health.Checker {
+	url := config.GetPortalURL()
+	timeout := 60 * time.Second
```

Comment:
The timeout and period may be different for each component. Do we need to expose them to the users?

Reply:
Suggest define them as constants with comments

Reply:
In addition to let `checker` define its own timeout, it would be good if there can be a general timeout, just in case some checker takes longer than expected to return the status and the connection of the API is lost.

Reply:
@steven-zou The `timeout` and `period` are the internal variables of the `Checker`, how to define them depends on the implementation of each `Checker`